### PR TITLE
Update verifyUser

### PR DESCRIPTION
**Pull request by AndrewJWin**
This PR updates ``./base/Command.js`` to be able to accurately grab the ``User`` object from either an @mention, an ID, or by Username or Tag.

Mention:
Mention

Username:
Username

Tag:
Tag

ID:
ID

### CODE REVIEW

File: base/Command.js [hunk 1]
```diff
@@ -53,10 +53,14 @@ class Command {
...
   async verifyUser(message, user, options = {}) {
-    const match = /(?:<@!?)?([0-9]{17,20})>?/gi.exec(user);
-    if (!match) throw new ParseError("Invalid Mention or ID", options.msg);
-    const id = match[1];
-    return this.client.users.fetch(id);
+    let match;
+    const idMatch = /(?:<@!?)?([0-9]{17,20})>?/gi.exec(user);
+    const nameMatch = /./gi.exec(user);
+    if (!idMatch && !nameMatch) throw new ParseError("Invalid Mention or ID", options.msg);
+    user.startsWith("<@") && user.endsWith(">") ? match = user.substr(2, 18) : match = user;
```
Comment (YorkAARGH) [inline base/Command.js]: This isn't needed? as the regex on line 57 will detect the user's ID from the mention anyway as you can see here 
Just grab (iirc) `match[1]` and you're golden.

**Review by a1icja — COMMENTED** [review]
Assuming that we are not doing speedwhoring, this looks fine. York's comment should be implemented before this PR is accepted.

File: base/Command.js (hunk 1 above)
Comment (AndrewJWin) [reply base/Command.js]: Oh, if thats the case that simplifies about 2 lines..

File: base/Command.js [hunk 2]
```diff
@@ -53,10 +53,14 @@ class Command {
   }
 
   async verifyUser(message, user, options = {}) {
-    const match = /(?:<@!?)?([0-9]{17,20})>?/gi.exec(user);
-    if (!match) throw new ParseError("Invalid Mention or ID", options.msg);
-    const id = match[1];
-    return this.client.users.fetch(id);
+    let match;
+    const idMatch = /(?:<@!?)?([0-9]{17,20})>?/gi.exec(user);
+    const nameMatch = /./gi.test(user);
```
Comment (kyranet) [inline base/Command.js]: This RegExp is unneeded.

File: base/Command.js [hunk 3]
```diff
@@ -53,10 +53,14 @@ class Command {
...
+    if (!idMatch && !nameMatch) throw new ParseError("Invalid Mention or ID", options.msg);
+    if (idMatch) match = idMatch[1];
```
Comment (kyranet) [inline base/Command.js]: After the RegExp check:

```javascript
if (idMatch) return this.client.users.fetch(idMatch[1]);
```

File: base/Command.js [hunk 4]
```diff
@@ -53,10 +53,14 @@ class Command {
...
 
   async verifyUser(message, user, options = {}) {
-    const match = /(?:<@!?)?([0-9]{17,20})>?/gi.exec(user);
-    if (!match) throw new ParseError("Invalid Mention or ID", options.msg);
-    const id = match[1];
-    return this.client.users.fetch(id);
+    let match;
+    const idMatch = /(?:<@!?)?([0-9]{17,20})>?/gi.exec(user);
+    const nameMatch = /./gi.test(user);
+    if (!idMatch && !nameMatch) throw new ParseError("Invalid Mention or ID", options.msg);
```
Comment (kyranet) [inline base/Command.js]: This should be at the end of the function and with no condition.

File: base/Command.js [hunk 5]
```diff
@@ -53,10 +53,14 @@ class Command {
...
+    if (idMatch) match = idMatch[1];
+    if (/(#[0-9]{4})$/.test(user)) match = user.substr(0, user.length-5);
```
Comment (kyranet) [inline base/Command.js]: Unlike usernames, there are only one user with the same username and discriminator (tag). You use this instead:

```javascript
if (/(#[0-9]{4})$/.test(user)) member = message.guild.members.find(member => member.user.tag === user);
```

File: base/Command.js [hunk 6]
```diff
@@ -53,10 +53,14 @@ class Command {
...
+    if (!idMatch) match = message.guild.members.find(m => m.user.username === match).id;
```
Comment (kyranet) [inline base/Command.js]: Then here, you try to find by username:

```javascript
// The previous check
else member = message.guild.members.find(member => member.user.username === match);
if (member) return member.user;
```

File: base/Command.js [hunk 7]
```diff
@@ -53,10 +53,14 @@ class Command {
...
+    return this.client.users.fetch(match);
```
Comment (kyranet) [inline base/Command.js]: With the code given above in the other requested changes, you'd cover:

1. Attempts to parse mention or ID
    - If success, return the output from the API (or cache, if the user is cached).
1. Checks if the input includes a discriminator
    - Finds a member by its user tag.
1. Definitively a username. Finds it without discriminator
1. If either the points 2 or 3 found a match, return [`GuildMember#user`](https://discord.js.org/#/docs/main/stable/class/GuildMember?scrollTo=user).
1. Else, there's no match by mention, tag or username. Throw the ParseError.

**Review by kyranet — CHANGES_REQUESTED** [review]
Could be improved a lot.

Bonus: Move the RegExp used for the ID match to avoid creating new instances of RegExp + the internal parsing.

**Review by kyranet — APPROVED** [review]

**Review by DevYukine — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by YorkAARGH** [comment]
[Update "Tagging" system to utilize user id, @mention and username.](https://trello.com/c/YuuMIgsj/9-update-tagging-system-to-utilize-user-id-mention-and-username)

**Pull request merged**